src/Http/Http.Results/src/ResultsOfTHelper.cs

```diff
@@ -11,6 +12,8 @@ internal static class ResultsOfTHelper
 {
     private static readonly MethodInfo PopulateMetadataMethod = typeof(ResultsOfTHelper).GetMethod(nameof(PopulateMetadata), BindingFlags.Static | BindingFlags.NonPublic)!;
 
+    [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "Only called from unsafe code.")]
+    [UnconditionalSuppressMessage("Trimmer", "IL2060", Justification = "Only called from unsafe code.")]
```

Comment:
It is not valid to be suppressing these warnings. The Justification isn't valid. It is possible for an application to get this method called through public API, and it would fail at runtime without getting a warning.

@jkotas @vitek-karas @MichalStrehovsky @agocke @davidwrighton  - The underlying issue here is that we need to conditionally/dynamically call a static abstract method, but only if the `TTarget` Type implements the interface. However, we can't make a generic constraint here, because it is called from a method that cannot have the generic constraint on it:

https://github.com/dotnet/aspnetcore/blob/c37ee916d1a17c1df0823e37ffdee458f3a12437/src/Http/Http.Results/src/ResultsOfT.Generated.cs#L64-L70

We can't make `Results<TResult1, TResult2>` have a generic constraint `where TResult1 : IEndpointMetadataProvider`. But we want it to "light up" when it does.

The way ASP.NET does this currently is below: checking for `IsAssignableFrom` and using `MakeGenericMethod(typeof(TTarget))` on a method that does have the generic constraint.

What is the recommended pattern for conditionally/dynamically invoking a static abstract method on a Type when a generic constraint isn't possible?

@JamesNK had a previous revision of this code using Reflection to look up the method, but it was reverted because it doesn't feel like this is the right approach:

https://github.com/dotnet/aspnetcore/pull/46082/commits/d7b1956a42e9213fcae968cdd037159fbe89dff0#diff-81c2e007d3867bb70fa117bc48dd221df618bf1790111f36a41a68cf73f4c251L31-L38

Reply:
One possibility could be to add a new API that gets a `Delegate` or a `MethodInfo` given a `System.Type implementationType`, `System.Type interfaceType`, and a `string methodName`.

Basically, we want to write this code similar to how it works with instances:

```C#
object obj = ...;
if (obj is IDisposable d)
{
    d.Dispose();
}
```

There is no way to do the above nicely for static abstract interface methods.

```C#
public interface IEndpointMetadataProvider
{
    static abstract void PopulateMetadata(MethodInfo method, EndpointBuilder builder);
}

Type type = ...;
if (type.IsAssignableTo(typeof(IEndPointMetadataProvider)))
{
    MethodInfo m = type.GetStaticInterfaceMethod(typeof(IEndPointMetadataProvider), "PopulateMetadata");
    m.Invoke(null, ...);
}
```

Reply:
That could make things simpler. I'm guessing we'd prefer it even if dynamic code is supported.

We'd still need DAM attributes to include methods because of reflection, but it would make `PopulateMetadataIfTargetIsIEndpointMetadataProvider` simpler.
```cs
public static void PopulateMetadataIfTargetIsIEndpointMetadataProvider<[DynamicallyAccessedMembers(RequireMethods)] TTarget>(MethodInfo method, EndpointBuilder builder)
{
    if (typeof(IEndpointMetadataProvider).IsAssignableFrom(typeof(TTarget)))
    {
        var parameters = new object[] { method, builder };
        var m = type.GetStaticInterfaceMethod(typeof(IEndPointMetadataProvider), "PopulateMetadata");
        m.Invoke(null, ull, BindingFlags.DoNotWrapExceptions, binder: null, parameters, culture: null);
    }
}
```

Reply:
I guess another option here is to not produce an AOT warning for this pattern:

```
private static readonly MethodInfo PopulateMetadataMethod = typeof(ResultsOfTHelper).GetMethod(nameof(PopulateMetadata), BindingFlags.Static | BindingFlags.NonPublic)!;

private static void PopulateMetadata<TTarget>(MethodInfo method, EndpointBuilder builder) where TTarget : IEndpointMetadataProvider {...}

...
PopulateMetadataMethod.MakeGenericMethod(typeof(TTarget))
```

The AOT compiler could be able to tell all the `TTarget` types that are passed into this method, and generate code for `private static void PopulateMetadata<TTarget>(MethodInfo method, EndpointBuilder builder) where TTarget : IEndpointMetadataProvider` for all `TTarget` struct types.

Reply:
> We can't make Results<TResult1, TResult2> have a generic constraint where TResult1 : IEndpointMetadataProvider. But we want it to "light up" when it does.

This sounds like https://github.com/dotnet/csharplang/discussions/6308 . There is no way to do this today without reflection.

Reply:
> We'd still need DAM attributes to include methods because of reflection, but it would make PopulateMetadataIfTargetIsIEndpointMetadataProvider simpler.

This would have to work without the DAM attribute - the problem is that the DAM wouldn't even cover all the cases - the static method could be default-implemented on an interface that `TTarget` has for example.

> I guess another option here is to not produce an AOT warning for this pattern:

The pattern would have to be easy to analyze (having to reverse engineer the value of `PopulateMetadataMethod` from the static constructor is not particularly easy). But doing intrinsic handling for `typeof(Foo).GetMethod("SomeStringLiteral).MakeGenericMethod(typeof(T))` sounds potentially doable. It is work, but doable. It would make the implementation a bit easier if we instead do `typeof(Foo<>).MakeGenericType(typeof(T)).GetMethod("SomeStringLiteral")`. The sole reason is that we already have facilities to do dataflow analysis with types, but we don't model methods.

> This sounds like https://github.com/dotnet/csharplang/discussions/6308 . There is no way to do this today without reflection.

That would be the best solution :)

Reply:
The reflection API solution is also discussed here: https://github.com/dotnet/runtime/issues/79331
Basically reflection doesn't have a "nice" way to invoke static abstract methods.

Reply:
Dataflow actually models MethodInfo a little bit already, but we don't have the ability to handle the static fields as a perf optimization. We could make `typeof(Foo).GetMethod("SomeStringLiteral).MakeGenericMethod(typeof(T))` work relatively easily. What is much more complex is to handle storing the method info in a static field. That work is basically this issue: https://github.com/dotnet/linker/issues/2482.

Reply:
For example:

```csharp
internal static class ResultsOfTHelper
{
    [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "MakeGenericType is safe because the target can be statically analyzed during compile.")]
    public static void PopulateMetadataIfTargetIsIEndpointMetadataProvider<TTarget>(MethodInfo method, EndpointBuilder builder)
    {
        if (typeof(IEndpointMetadataProvider).IsAssignableFrom(typeof(TTarget)))
        {
            var methodInfo = typeof(MetadataPopulator<>).MakeGenericType(typeof(TTarget)).GetMethod("PopulateMetadata")!;
            methodInfo.Invoke(null, new object[] { method, builder });
        }
    }

    internal static class MetadataPopulator<TTarget> where TTarget : IEndpointMetadataProvider
    {
        public static void PopulateMetadata(MethodInfo method, EndpointBuilder builder)
        {
            TTarget.PopulateMetadata(method, builder);
        }
    }
}
```

Reply:
> This sounds like https://github.com/dotnet/csharplang/discussions/6308 . There is no way to do this today without reflection.

> That would be the best solution :)

Assuming that isn't going to make it into C# 12, given that it isn't listed on https://github.com/dotnet/roslyn/blob/main/docs/Language%20Feature%20Status.md#c-next, what is the recommendation for a solution to this problem in .NET 8?

Note, we can't make this method `[RequiresDynamicCode]` because it is called from an implementation of the `IEndpointMetadataProvider` interface. Which means that we would need to mark `IEndpointMetadataProvider` as `RequiresDynamicCode`, which is not desirable/possible.

The only viable options for .NET 8 I can see are:

1. Push the language team to solve https://github.com/dotnet/csharplang/discussions/6308. (I'm not sure how "viable" this is)
2. Add a Reflection API, for example `MethodInfo Type.GetStaticInterfaceMethod(MethodInfo)`. This can be called to get the implementation of a static abstract method.
3. Add the above pattern/heuristic into the AOT compiler and use it in the ASP.NET code.
4. ASP.NET do it's own implementation of the Reflection API in (2), for now until an official solution is created.  (This is @JamesNK's earlier proposal on this PR.)

Reply:
I tested my sample above and it fails today in AOT with a value-type argument.
I also tested the reverted reflection API solution in AOT and it works.

The only workable solution today is 4. We can either use that now or leave `Results<...>.PopulateMetadata` it broken.

@davidfowl Are you ok with using reflection for now, and eventually replacing it when a better solution is available?

Reply:
> I tested my sample [above](https://github.com/dotnet/aspnetcore/pull/46082#discussion_r1086689999) and it fails today in AOT with a value-type argument.

This would be fixable in the compiler - we could intrinsically detect this. I'll look into it.

It was brought up in the past but I couldn't see a use case: https://github.com/dotnet/runtimelab/issues/1641. Now the use case is clear - we'd use this for "bridging generic constraints" where we really want a static reference, but we're disallowed from doing that due to constraints mismatch.

Reply:
Is there an existing issue for this? If not, where should I create a new one?

Reply:
https://github.com/dotnet/runtime/issues/81204